import os,glob,sys

# /BDATA/smkim/imputation.tool.check/impute5/OUTPUTs/Ref.KRG1KGP.haptovcf_eagle.vcftovcf.phased/bcftools.stats.test
#bcftools stats -f 'MAF>=0.01 && INFO >= 0.8' maf0.01.info.0.8.filter.Tool.imputation.merge.processing.chr1.247000001_249240543.vcf.gz
#ls {} | xargs -P 32 -I{} bash -c "bcftools stats -f 'MAF>=0.01 && INFO >= 0.8' {} | grep -v "^#" | head > {}

#ID	0	maf0.01.info.0.8.filter.Tool.imputation.merge.processing.chr1.247000001_249240543.vcf.gz
#SN	0	number of samples:	10000
#SN	0	number of records:	5786
#SN	0	number of no-ALTs:	0
#SN	0	number of SNPs:	5195
#SN	0	number of MNPs:	0
#SN	0	number of indels:	585
#SN	0	number of others:	6
#SN	0	number of multiallelic sites:	0
#SN	0	number of multiallelic SNP sites:	0
#"10178_5000000","5000001_10000000","10000001_15000000","15000001_20000000","20000001_25000000","25000001_30000000","30000001_35000000","35000001_40000000","40000001_45000000","45000001_50000000","50000001_55000000","55000001_60000000","60000001_65000000","65000001_70000000","70000001_75000000","75000001_80000000","80000001_85000000","85000001_90000000","90000001_95000000","95000001_100000000","100000001_105000000","105000001_110000000","110000001_115000000","115000001_120000000","117000001_121485368","142535440_147000000","147000001_152000000","152000001_157000000","157000001_162000000","162000001_167000000","167000001_172000000","172000001_177000000","177000001_182000000","182000001_187000000","187000001_192000000","192000001_197000000","197000001_202000000","202000001_207000000","207000001_212000000","212000001_217000000","217000001_222000000","222000001_227000000","227000001_232000000","232000001_237000000","237000001_242000000","242000001_247000000","245000001_249240543"


#wDir = "/BDATA/smkim/imputation.tool.check/impute5/OUTPUTs/Ref.KRG1KGP.haptovcf_eagle.vcftovcf.phased/"
wDir = "/ADATA/smkim/imputation.tool/impute5/shapeit4_usingKRG1KGP.haptovcf.ref/"
file_lists = glob.glob(wDir+"*gz")
chunks = ["10178_5000000","5000001_10000000","10000001_15000000","15000001_20000000","20000001_25000000","25000001_30000000","30000001_35000000","35000001_40000000","40000001_45000000","45000001_50000000","50000001_55000000","55000001_60000000","60000001_65000000","65000001_70000000","70000001_75000000","75000001_80000000","80000001_85000000","85000001_90000000","90000001_95000000","95000001_100000000","100000001_105000000","105000001_110000000","110000001_115000000","115000001_120000000","117000001_121485368","142535440_147000000","147000001_152000000","152000001_157000000","157000001_162000000","162000001_167000000","167000001_172000000","172000001_177000000","177000001_182000000","182000001_187000000","187000001_192000000","192000001_197000000","197000001_202000000","202000001_207000000","207000001_212000000","212000001_217000000","217000001_222000000","222000001_227000000","227000001_232000000","232000001_237000000","237000001_242000000","242000001_247000000","245000001_249240543"]

out = open("%sfilter.info.txt"%wDir,"w")
out.write("chunk\tsample\trecords\tno_ALTs\tSNPs\tMNPs\tindels\tothers\tmultiallelic_sites\tmultiallelic_SNP_sites\n")
count = 0
for i in file_lists:
	count = count + 1
	print("Count : %s\t%s"%(str(count),i))
	os.system("bcftools stats -f 'MAF>=0.01 && INFO>=0.8' %s | grep -v \"^#\" | head > temp.filter.txt"%(i))
	tmp = open("temp.filter.txt","r")
	tmp1 = [a.replace("\n","") for a in tmp]
	for row_index,j in enumerate(tmp1):
		if row_index == 0:
			for chunk in chunks :
				if chunk in j:
					out.write("chr1_%s\t"%chunk)
					break
		else:
			out.write("%s\t"%(j.split("\t")[3]))
	out.write("\n")
out.close()