{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manhattan plot\n",
    "#### R programing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(qqman)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>SNP</th><th scope=col>CHR</th><th scope=col>BP</th><th scope=col>P</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>rs1      </td><td>1        </td><td>1        </td><td>0.9148060</td></tr>\n",
       "\t<tr><td>rs2      </td><td>1        </td><td>2        </td><td>0.9370754</td></tr>\n",
       "\t<tr><td>rs3      </td><td>1        </td><td>3        </td><td>0.2861395</td></tr>\n",
       "\t<tr><td>rs4      </td><td>1        </td><td>4        </td><td>0.8304476</td></tr>\n",
       "\t<tr><td>rs5      </td><td>1        </td><td>5        </td><td>0.6417455</td></tr>\n",
       "\t<tr><td>rs6      </td><td>1        </td><td>6        </td><td>0.5190959</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       " SNP & CHR & BP & P\\\\\n",
       "\\hline\n",
       "\t rs1       & 1         & 1         & 0.9148060\\\\\n",
       "\t rs2       & 1         & 2         & 0.9370754\\\\\n",
       "\t rs3       & 1         & 3         & 0.2861395\\\\\n",
       "\t rs4       & 1         & 4         & 0.8304476\\\\\n",
       "\t rs5       & 1         & 5         & 0.6417455\\\\\n",
       "\t rs6       & 1         & 6         & 0.5190959\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| SNP | CHR | BP | P |\n",
       "|---|---|---|---|\n",
       "| rs1       | 1         | 1         | 0.9148060 |\n",
       "| rs2       | 1         | 2         | 0.9370754 |\n",
       "| rs3       | 1         | 3         | 0.2861395 |\n",
       "| rs4       | 1         | 4         | 0.8304476 |\n",
       "| rs5       | 1         | 5         | 0.6417455 |\n",
       "| rs6       | 1         | 6         | 0.5190959 |\n",
       "\n"
      ],
      "text/plain": [
       "  SNP CHR BP P        \n",
       "1 rs1 1   1  0.9148060\n",
       "2 rs2 1   2  0.9370754\n",
       "3 rs3 1   3  0.2861395\n",
       "4 rs4 1   4  0.8304476\n",
       "5 rs5 1   5  0.6417455\n",
       "6 rs6 1   6  0.5190959"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gwasResults)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAPFBMVEUAAAAAAP8aGhpNTU1o\naGh8fHyMjIyZmZmampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD/AAD///9+Jsh2AAAA\nCXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2di3abPLdFdeymbpKvafLz/u96EpuL7hfY\nQsLMOUYTKgkh9mKJDcZEDQCwGdV6AADPAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwE\nIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATAS\ngAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBI\nAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAj\nAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACM\nBCAARgIQACMBCICRAATASAACYCQAATASgADtjXS5tB4BpECjJM2NdLmgUu+gURqMBEnQKE1z\nI5E2HAA0SlJiJPWN/AgQSRI0akRB1FVh+zxIGyRBo1ZgpKcCjVqx3khqZOsI0EgQNGrF5muk\n9ncrQAONGrE5tUOknkCjVmCkpwKNWtHeSOTfgqBRK5obiTtCkqBRK5rfbEAkUdCoEZvP+ojU\nP2hUH4x0AtCoPhjpBKBRfZobiTtC9UGj+myO8f9gKxI6olFdkjFufkYibagPGtWnuZFIG+qD\nRvVpbiRmu/qgUX0w0glAo/pgpBOARvXBSCcAjeqDkU4AGtUHI50ANKoPRjoBaFQfjHQC0Kg+\nGOkEoFF9MNIJQKP6YKQTgEb1wUgnAI3q83hbxob3cSLSPqBR36gfiablRMB5sUYz0Kh3lBHl\naMR5Z1oz0Kh7rPkrFnJ/HSLVB436p/QFkdp/Zf7SAWmDJGjUiiIjkX/3Dhq1YoxxzrR1b8Hr\ncJuBRj2jlp+pcPNe6aagUdco52e8KSK1Ao26Rmm/EKlr0KhnxnnuJ7fOSMC5kG0HGnXNHOO1\nd0kRaT/QqF+MGK/RCZH2BY36hKe/TwAa1afgEaFABxsHgEZp0Kh/Ch5aDXWwCSa7HNCoe0q+\nRhHqYAuIlAUa9Q5f7DsIaNQ33Gw4AWhUn81G+j/YioSOUbjZUJ8lxiszB/LvHUGjblHmwu63\nVi+olA8a9QtGOhBo1C9WatdEpG099MLtVn8baLSNihp1cNdu6wj64Hbbw0nrQKMHNTVqbaSn\nme2e2EholIHxNQrShg3s4CM02sguqd3KaJ8gbejnVINGIZpr1Dq1OwA9J215oFF9MFKS9iJt\nBY3qM7+zYe0zkaQNO4FGMZprpL1FSA3pkIu/WON5LmSrg0Y9o8yfqbc9VREJlXJAo65pbSSm\nu1zQqGum1E7lvDbN+F660F86OEIC3gdo1DNFMVby+TciCYNGbSiJsXc6JG3oCjRqhJnaJdsi\nUjvQqGeU8zPc1KvkUrAy2GiUBxp1TYGR/E3mEqatuqBR12CkPdkSITTah5URKrlGinWgjyA8\nlLwHOZo/7lGLbccxGu3BWo0qPLQaHkreo4V7PoC47wzd6oSARkVbO6aR3MY7irT3wy+NEis0\nKtreqtWU9ljxKlN5VipJG3xhKtVofZwP8xQZGvVO6+8jFYRpaWequCXQO2jUwdUEGiUQ0Ki1\nkfLDtGhh5RXe+bLd0WuNpv1XztDIoYJGzY2UTVAkN224NTx67elbG0qzDAWNTGpodAgjPfYu\nlDbobab6miLFu3byIE2jVk5CI3sc8hopZ2FtB24sB89/15Czd2YbQ6OxYqtsobnWM5JgxcZY\noFFyC3Pfe2skaKTHIKyhaOb/2bdVkXL37tGNqUQiNpHYzhU/C/Fuvv+tn0k3H69o1K9Gcka6\nhEW6F9xGVmzD1eghuGdO862cEGmu+VkIin2vuP/YOG1uWB2N+tVI1kiP30b9apGsuz5mlStS\nmFTa4BfJ0cpzBK5hy6UBGvWrkXRq59mJdWmDsTP+i8NIZ8ntmNn6vMlxQ8sc7V/tvsY6QzQ1\nEhpl0YGRhO5Mfe+HT6TsvpP3M/3jNDUKb+zevS/a95LEKFumdmhUTSPJ299jgurNHoJD8wbr\nEQS7lT90wQ7GWt8U7O3pe5XUZDc2C4l0LxI6TiVBI2uzFTSSM9KkzWOQ1lC1HNdc3btLU2ur\nsXcS0kuWJeODAWclb6Ee4fGX78haDiDfZOeItCWTEwON/F1Yu7IJMSPNY/OItFzB2mMOzI7+\nxmNrTw/hy8zLdB9H7ztTJG+Ao1HX0gb96j3YfifQyB63vEbyRtIH6YxUH/OcZQQC7N2/H420\n0imul6UnI+NYFNR79Uuqz1TjZOef7twxua3m+dLtZH9roVF9jaRTO415Rrl5VbpoeO/5GDsz\n1xp7PUXc0Mg8EKyJbjC2b+/MMvUG0wYXu6P5qJlHaOxCg5MUGtXXqOi9dr6vw+ifURgVc9g8\n093PD02fwbnYHFtN6cASCHMPx//MGo1dGyIZjR8qecc7Tnfz2M1pM4Il0rhb3mhYwagAGnmp\nr1GBkZS3vZ42aOPS5p9pZPPSOM7QVDc1MLrRRRr0ttP6D41ujkbata1R7QzOiN3lYqptaG2H\nwCuS2dzehVpGQqNpeBbVNSpN7TJFCsV2uUK9l8wzg5Ote3KLy+ViiD03mJvNM+kS2/miQNdo\nmSEHU7ybq6nVj5sHmMei0XjJfebD8XYbnCNNGjRqoJGYkUzPmyLNe6v/vj2ygjHIy87pwVly\nWV0GvadZ37F66X0exmPBFmmqszQyxXBFutjK6YePpoi2l8vBOGiHUWHcS0CjBhqtN1L0JQK6\nRpdp9Mvvxy7M7QZjSRPJ+hj7LoMejzlj9oikaziM+s6DWGbXuTNNpMt0FTutshwqTt8Xo/ze\n81RhaG8V7WMkNNpNo0IjRS5kJy7LHtk7sITJaL3IME4vU4vbzVbJ7Um79LzoMtyXjUPA7fVe\n85B96cS8m6MF2RiAtaeD3tQZ7SLSHk5CoxYalRnJ09oomvclpNK068vloCGCJaMezoBIerj1\n/NsM7mMCu80fd2trPVazhm8l5r59MZsbTR2RDPVXiFQEGjXRqMhIvsZ6mbaPhkjmmV4fpdls\nPilr2foStHHKMvqa66bpzquRMYhlq8Ok3XT4GK0DPc1bnzZuaGRudAiU1bzdgEZtNCr6HClV\nGNJIE0bbHVej+b7NsCTRTpc+jZ3ryWki9IxisGoGf/ngTHfjBsfr29syZ5qTsNuzo/FQz0lo\n1Eijks+REn8yxA65RyOTcNPAKdvYc6PSuZ60pdBnKavKPyJ3O/pFwlT42G3/mOwL2YulZg0j\noVErjSQfEdJ3MUOksEaxVvmV3rR51D/dYbonfx/a7id63Rp5NOpJI9GHVheNjGtC435oKMJ2\nHlssw1TrfJZoiWRnFIZ6/g0FBpAo9Ii0TqPNoFF9jeoYySUpkk+DpEh2daQyVBNcI9lPso+o\nSFvjjkYBMQLFtTWqlNrZsfNo5BXJaefpK1ydWxdcJ6WGf14bAn24F7KzQqUyoVHvGgl+1dyr\njxu7mEZO01SDWGW4KlgTqPC3TvcRb7ifk9Ao3IeURrLvbPAjJtKQqI9UmtEJ1IR6S7aOl4bG\nsTX0aNSRRjuldrkqpRqViJRZlbFGunW8aWAYW0OPRh1ptM/NhtUaxWMd1zCyYqgmUO4vztcj\nMIytoUejjjTaxUi+2OeKFG0QXX9FTVGxtzRe1rGR0Mg7jOIY1xTJq5FXpWSjktpwVagiXe7v\nJtE0NIytoUejROBvO2p0MCPFdDiASOYAt4YejRKBv+2oUbvUziNSulFBbWZVqCJdnF0YGv7W\n0KNRsji7MDT84hhvFql4ustrF28Rq82syihPFmeL1Di1Q6N6GnV1syGjUb5IcmlDujgukje/\nKBEJjfrX6GBGild3KlJogFtDj0YdadTuZkNWw0SLaG24KlSRUZ4qjZc1Tu184gRDj0ZFGmGk\ndHm6OFckc4BbQ49GHWl0bCPlXq2KiVRQGi/7YWvo0ShfjeoadWUkt2WiQaw2s2pXkZ4gtUOj\neIz3F8mjUrpNgUgr0oaVcjzvGQmNimOc1Tb2lw4CAvlj74u/t2GiRbQ2XBWqCJT7i1NtBUVC\no/41kv1rFH6eUyR/7H2FoQHmhx6N8sVopNHzGimz6ggioVH/GmGktSKl2voKV4qERv1rtN5I\n1l868KkTDr5fpGKV8ivzalKBFxUpP/Jo1L9Gm89Ipf0m36kbKLELVKxWlf8nYzm1mCj0NhQG\njRKL1TTCSPnLGClRcmaNMFL+MkZKlJxZI4yUv4yREiVn1qikffCPKJb0i0g1jYRGicUujCTS\nLyJVNVIJaJS1lAdGyl/GSImSM2uEkfKXMVKi5MwadaIpwLHBSAACYCQAATASgAAYCUAAjAQg\nAEYCEAAjAQiAkQAEwEgAAlQyUqBb/8uiYm+QKugnWamSq2Y1SrWI1Y/FJc9oVwONJDWqo2c0\nSlmFGYEsrXyEJ7ZqVqNUi1j9WJwaxC6gkb9+pUZV5FTRMOQUZogUrfS0eAwqER9t5KEJe+km\nVR/sIdb/XqCRrEb7pnbBKmX/V16kIUekISnSkBApvZk+jIRGz2ckNxlNirQ6/84UKZV/x0Yw\nbSbU4JBGQqMEuxspHD/jf3GRYhPOPrNdZASPTajQgdaLj9BIVKO9jRSfAZb/rcwMopXSaUNq\nX2IJUnsfoVF0EIkhhHuUJRrBVLnxdtBg62OK1I+P0ChUv0qjfY0UHLZT0XS2izfKESmUVXTk\nIzQKVK/TaFcj+aexZ/2wLzA0pdLT+S6gkaRGzeUEeAYwEoAAGAlAAIwEIABGAhAAIwEIgJEA\nBMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYC\nEAAjAQhwbCONL0w69k48OSfR6ND7F3vNH/TBWTQ68v6p6eeRd+LJOY1GR94/7QXO418eUFMm\n8ShR+gs1jV+eeqjBaTTqfXwxFpEeM974a15U81To/nLqoQqn0ajz4UVR+oIWbu+iSjWFGpxG\no86HF6VIpDlfcOp7eKH903IajTofXhQ1/cwSKdYUanEajQ4wxDBqSCnjyHLItOHInEWjzoeX\nQPuwb5q3jOxAyxfsO0JOPdThJBr1Pj6AQ4CRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAj\nAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACM\nBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEw\nEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATA\nSAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAA\nIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAA\njAQgAEYCEAAjAQiQaaR7M/VNzbHAJtCoJXlRv4uj8tvD/qBRU7KCrgZE6h00aktBameJpEZq\njApWENbof7CVrOivE6msA6gOGrUEIz0NaNQSjPQ0oFFLMNLTgEYtwUhPAxq1ZPMHsojUC2jU\nks0xRqT+QaP6YKQTgEb1wUgnAI3qg5FOABrVByOdADSqD0Y6AWhUH4x0AtCoPhjpBKBRfTDS\nCUCj+mCkE4BG9cFIJwCN6oORTgAa1QcjnQA0qg9GOgFoVJ/2RrpctvYACdCoPs2NdLmgUm3Q\nqD4Y6QSgUX1KYlzn25doJAkaNaIgxnXeB8BsJwkatQIjPRVo1Ir1RpJ6ZTEaCYJGrWh/jQSS\noFEjmqd2IAkatQIjPRVo1Ir2RiL/FgSNWtHcSNwRkgSNWtH8ZgMiiYJGjeARoROARvXBSCcA\njerT3EhcyNYHjeqzOcat/0ruEyChYxTOSPVpfkZCpPqgUX2aG4m0oT5oVJ/mRmK2qw8a1efx\nCd6GZ4QRaR/QqG/Uj0TT8pqAI9IeoFHvKCPKKyKOSDuARt1jzXDlIUek+qBR/3Cz4QSgUX0w\n0glAo/qMMV7/xX5E2gs06hm1/OTWategUdco5+eqDtaDSJmgUdco7RcidQ0a9cw4z/1k38lw\n8+3LlqBR18wxTl/K8mKN1qBRvxgxjuvkr0OkfUGjPil9i5D2X5nX4ZI2SIJGrSh4REh5Z0NE\nqg8a9U/BQ6t3hXhnWgvQqHsKvkZR50IWkXJAo94p+GIfIrUEjfqm9GYDInUNGrWiJMZ82Nc/\naNSIJcYr75Mi0o6gUbcoc4FvX/YMGvVLD0ba2MN5QKN+sVK73UVisisBjbpl82NYiNQ/aFSf\n1kZCpR1Ao/oYX6NokDYgUgFo1C/Ks1TE/8FW1qiFRvtSSxqxDi7MdvVBo/pgpBOARvWZ39mw\n9htg3BEaud3q9o9G26mokVp+eb/KktvBap5Ho+pOGtBoGzU1UuZPPuxbS3UjodFmMNIR2OOE\nhEbb2CG1U3mvTQt2sJoDaFTbIpmgUYTmGrW+a3cA6l/91AaN6tPcSP3Pdu1F2goa1cdM7XKb\nR0uKOEL+3VqjETSK0Fwj5fyMtq4iUvcqdQEadU1rIx1iuusCNOqaIiMZb5MUeh3uARLwPkCj\nrim6RlLy+TciZYNGPVMSY++HGKQNXYFGjRA00rpgI5IoaNQIpT1WnDCVv81csDbaaJQBGvVO\n4Vk/MtsxbaXZJUJotImVEZIz0jKC8FDyPjZr/uFaLfY5jtFoC2s1qvCIUHgoeQ9y7Pm4x74z\ndKsTAhoVbe2YRnIbF4u0PtB7f2bfKLFCo6LtrVpNOQtrO8gZilcjj0relUPdbgj0wR5+QaN+\nqWGkEvLDpDU0VcyX2dtpZsvVCOZAaFQJAY1aGyk/TIsWdl6xPfGQxNohybGgkRAVNGpupHy0\nyS6x37eqKsW7tqdvbSibZ1Y0ymV/jSSN5B2ExHnZ7sMXplBkxK7wQ3Ot08yZ7rSabUNAo8xB\nNNBI0EjjIMyRXPSz/cpNuHv36Erv0JljnJrw5uea2Agf3Xz/S82kQSl6MBIaTQ1jPaxB7vb3\nZdoNYyiXyxKk751bJZSzd484WVNaIjbh2M4194VoN/cfGw62x1ZWrr4eNMpmtUayRvKcNQ2R\nSvLiiAIekSLJgXaMBDakiTRvyunPOQLXidXiIhuNylijkeAHspcJq4WWNuSO0G5qZyRu2uB2\nEC/SRznVjBpd9DnaYix0j5FsmhoJjbLoxEix1r5P+3ytrInRm5FEN2PchnFX83dlahTe1pgB\nuTtzL0kMsmVqh0bVNJJP7cZBOimzf3VvMNxUfZzsfK21Es/NV/+R4+3pnnzHJ7ux2ZS2eMdt\nrtriBOSARta4K2gkZqR5bN6ZSctxzdX94ZjSAl/ji33+nzMKrSvjfqYdNr9yeoTHX774xq7H\nzS6mkvZOQqNQF1PJdo12MdIyVHvIS7RcqXz7d9dIK50y5os37sN4o9OefqIiLfvgja93UEvl\n9N9Z5qMYCY22IZnaGQuO5WeRtMx4iZYds/HkPO/fVHsxSvWIX+a4DPpq9hXrqIUnYdFP+XOE\n3R32HThOVmKIZO3C/tZCo/oaFb2zwfdNZz3/todqWH65fXmP/yLQLJa2rtZq7ktb32imaTRJ\na2QNeuPbkoy4sV1urN5/505UdkfzXj36tIJR+SSFRl7qa1RgJOVtr6cN2lj1+ceYgG4TWgtT\nI32i01WcV9e2/tPNsrHbwjKKwd6w3rXW5zioeT2rI22TVgQskawDbtnYHkZCo2l4JvU1Kk3t\nMkXSZqBhGtk45HlGWma5sZmRZNzsfu6Nlv0z5ixNpMGMrZluaH3PgzMq5uYXral1iWzlAboo\nP1Xmf62tXawBVgGNGmgkZiQzbTBFmof1+KXFahq7c+13c/pZZNBnjFnfsXpwasdIGlrclrBZ\nGpliTFubyy9a/m8cH4PexcU83G66RvaFRQXQqIFG640UfT+ULtJlisv4e/mhTxVLlDWRHtGx\nRVoiPSfoHpGM3OAyR3cKpXVM3HudBuPMgssonL4vRrl2s0k7LjxFNZ2ERg00EnyL0Lh3xtiN\nHdDCpK1h7+qgJxnTPOPrarohtPSjy3Bfno8BrXNj9hqWQ2jeii2fcdNJD7F2BBlNndEuImk7\nUBb4AtCohUZlRvK01ouWfTHP93Zg7bCYDYc5ub5pZ+G5hTHZTSI8+rqZaB0PjwxBv6wc1xls\n8WZhvR3pKi176t+oPQU+VL2vvKuP0GgHjYqM5GvsF8mMjyaStjvuCsbUY944XdpqWYGe7BpR\nuWn6XqwVl80OZrG5RkSlQS+15jDz0AgI99i5ksijUe8aFb77O1poCGONyyOSE8WlbSA++o7f\nw7pE1NqUpqA9CEc//7bckU67aGkUFsm+uJ03Zn/UIggaNdKo5HOk+HulDWHC+XeORrFW8S70\nSv8YBmdFf3/BUaQaL7vvaWl0mh96NOpeI+FvyHpFuliznLM3azXy1mo3eYIi+Xq0KkKb8Rb7\n27pl6zTaDBrV10j0odUIHpU8rUpFsqsjlaGa4BrJfvx9+Jr6Gm6NOxoFxAgU19ZoFyN9D9MN\nf17akGixri64TkqN9GTnLfY33Bp5NOpJo0qpXTzyQY2cpqkGscpwVagm1FuydaDzeMP9nIRG\nl+oaHcpIQ6I+UplVE+psY2loHFtDj0YdabRTapcpUrJViUiZVRnlBaWJQmMUW0OPRh1p1LuR\n4vNZVMMVNYHygtJ42XMaCY2GfYzk1WjdhWyii40ihVZIdZNoGhrG1tCjUSLwtx016stIyUYl\nteGqtBh9iYRGJeX+bhJNQ8MojnFNkbKvZNON4tVrqgrLE3okRDIHvzX0aOQvT8hRR6M97tpl\n59/pVgW1mVUZ5cni2iKhUf8adXWzIaNRvkhyaUO6OCFIYIBbQ49GHWnU7hopq2GiRbQ2XBWq\nSKux8Y7QSpHQqH+NMFJZeao0Xna0mw1ZDRMtorXhqgwtutLoeY2UWZVRnizljBRpEavNrDqC\nRhhJQCRvqaBIaNS/RhhpT5FI7Z5WI7FXFntiHo79HiJVuZDdKJI5wILQo1GifF1pvOyH/MAX\naORtfy6RknNYvOVKkdCof42e10iZVatF8sfeVxgafn7o0ah7jTDSWpFSbX2FK0VCo/41Wm8k\n63W4PnXCwfeLVKxSfmVeTSrwoiLlRx6N+tdo8xmptN/Ui9dCJXaBitWq8v9kLKcWE4XehsKg\nUWKxmkYYKX8ZIyVKzqwRRspfxkiJkjNrhJHylzFSouTMGpW0D/6lnZJ+EammkdAosdiFkUT6\nRaSqRioBjbKW8sBI+csYKVFyZo0wUv4yRkqUnFmjTjQFODYYCUAAjAQgAEYCEAAjAQiAkQAE\nwEgAAmAkAAEwEoAAGAlAgEpGCnTrf1lU7A1SBf0kK1Vy1axGqRax+rG45BntaqCRpEZ19IxG\nKaswI5CllY/wxFbNapRqEasfi1OD2AU08tev1KiKnCoahpzCDJGilZ4Wj0El4qONPDRhL92k\n6oM9xPrfCzSS1Wjf1C5Ypez/yos05Ig0JEUaEiKlN9OHkdDo+YzkJqNJkVbn35kipfLv2Aim\nzYQaHNJIaJRgdyOF42f8Ly5SbMLZZ7aLjOCxCRU60HrxERqJarS3keIzwPK/lZlBtFI6bUjt\nSyxBau8jNIoOIjGEcI+yRCOYKjfeDhpsfUyR+vERGoXqV2m0r5GCw3Yqms528UY5IoWyio58\nhEaB6nUa7Wok/zT2rB/2BYamVHo63wU0ktSouZwAzwBGAhAAIwEIgJEABMBIAAJgJAABMBKA\nABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgA\nAmAkAAGObaTxhUnH3okn5yQaHXr/Yq/5gz44i0ZH3j81/TzyTjw5p9HoyPunvcB5/MsDasok\nHiVKf6Gm8ctTDzU4jUa9jy/GItJjxht/zYtqngrdX049VOE0GnU+vChKX9DC7V1UqaZQg9No\n1PnwohSJNOcLTn0PL7R/Wk6jUefDi6Kmn1kixZpCLU6j0QGGGEYNKWUcWQ6ZNhyZs2jU+fAS\naB/2TfOWkR1o+YJ9R8iphzqcRKPexwdwCDASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAk\nAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICR\nAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABG\nAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAY\nCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJg\nJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEY6OO+zgl+v\nv5R6eftqOZrzgpGOzaeaFPy8qjvXz6YDOisY6dD8uGdc/K1ePobh40X9bjqis4KRjsz9LDQu\nql+PhasiuWsARjow70q9TkZ6V2+PhTf1X7sRnZdMI92bqXn6gy5Q6n2YJPmj/j0WPtSfdiM6\nL3nOuKul8tvDLvz+HGYjvajxJsOnemk3ovOSZQw1YKRemYy0JAukDS0oSO0sIz1utqr/wVa2\nCRgzEhpJkeWQHLmG0BmJ6a8tOWckNKoPRjo4GKkPMNLBwUh9gJEOzmSb39PnsF/uXTs0qg9G\nOjjO50j/3M+R0Kg+mz+QRaS2TJL8tzzZ8O602XNAJ2VzjBGpLcuzdtfHwlU5j3+jUX0w0sGx\nn/7+5Xn6G43qg3OqAeQAABKTSURBVJEOzmykf+HvI6FRfTDSwVkuW+/fkP3l+4YsGtUHI50A\nNKoPRjoBaFQfjHQC0Kg+GOkEoFF9MNIJQKP6YKQTgEb1wUgnAI3qg5FOABrVByM9D39flHr5\n66lAo/pgpKfh9fGI0Ktbg0b1wUjPwl/18m8Y/r0o95yERvVpb6TLZWsP8MOL+vj59eF5rR0a\n1QcjPQvuyxuWqq19o1GSkhjzDdmeqWkkSFIQ40rvbGC2k4HUrikY6Vn4uN9s+Bz9ZIBG9Vlv\npPF1uKQNvfD3/hXZq37TDo12g2ukZ+Hz98M0nj/qgkb1aZ/agQhfV/X+9W2n9+ltQhpoVB+M\n9CS8Tq+ze3efbUCj+mCkJ+G6vLLYOSWhUX3aG4k7QiJUfYk+GiVpf7MBkUS4TkLUOCOhURIe\nEXoSXqc/Zl7jGgmNkmCkJ+Hrqt4+v3+9q6vzikg0qk97I4EMf6dXFvM1iha0NxKznRCPVxa/\n1nhlMRolwUgnAI3q095IUB00qk97IzHbVQeN6vP4lHXDM8KItA9o1DdK+yR8lUyItAdo1DvK\niPKKiCPSDqBR91gzXHnIEak+aNQ/3Gw4AWhUH4z0PHy+XpX688+tQKP6zO9wWvvFfkTai5RG\nf8cbexUeEUKjJGr5ya3Vrklp9E9d//savt6Ucs5JaFQf5fxc1cEGECmPlEa/xu+av7mvP0Gj\n+ijtF0bqmrhGH9P3+fhiXxPGee4n+06Gm2/ItiSh0Zt6C6+6ddtolET7pn/qdgMvP2lNTKPf\n7qXRsl6V0YCOEeO4l/x1iLQvIY1+qa+v16u6Vvk+EiQpfYuQ9l+h1+GSNoig1OfjK7LXD70Q\njXai4BEh5Z0NEak+ORp9O+jlg3c2NKPgodW7i3hnWgsyNFLq92PhjbcItaDgaxSVbjYgUgZp\njb5Tu8fCp/rl1G3dPBolKfhiH0ZqSUqjF+3+q7Pu1o2jUZLSmw2I1Cl/MFJTSmLMB7Id89/0\nsOqn+7cv0ag+S4xX3idFpB2JaPQ1XRrNf99FW23rdtEoiTIX+PZlz0Q1elMvP68sfqvxh8bQ\nKEl7I0E2cY1ex1cWfwbWg5pYqd3+RmKyKyCu0cefSq8sRqM0zb9qjkj1QaP6NDcSKtUHjepj\nfI2iRWqHSPmgUb8oz1IR/wdbWaMWGu1LLWnkOrgw21UHjeqDkU4AGtVnfmfD2m+AcUdoJ9Co\na7S3CHm/bpTbAdxulTeARpupqJEyf/KB7GpqGylPow/fGQuNRjASZGr0CyO1YUrtVN6r7YId\nQHVyNHr3XkOhUX2a37U7ANWvfsT4VNeTGqm5RhgpTXORsvmt/sNIbTBTu9zm0ZIyyL9zSWv0\nV714346CRvVRzs9o6xpGQqUsMjS6qs9KRkKjFM2NxHSXSVqj15/36FdJ7dAoSZGRjLdJSr0O\nF/JIavTv/iVzQxA02o2iayRV4RoJcklp9KI+hkpnJEhSEmPvhxikDZ3w/nhnMaldGzDSk/B1\nVfe3NWCkNijtseJEvP1tlgKiXY+0Rn/G19nFjYRGtSicrGJnJERqiVJhs6FRfQSNBEkqHsd5\nRoIkKzWq8YhQeCh5D3I0f9yjGjucEDKvkdAoBEZaw/N9Zv98RjqGRspZWNvBSgTCtKGHY4g0\nkfU0ZJ+PCD29RgcyUrDhIQItQf9GOq9GzY2Uz6JFMq/oKYMXHAsaVUJgLAcy0kKGSO1Usqfv\nUxlp4WwaHdJIaWqKlOg7nAdtTm/QSKjvChpJGsk7CInkOKePcBuh9HzuZvUB0IOR0Cizh1IE\nb3+PNreGcpndv34Gcvbu0ZXRYVKk8ObnmugI790c85oZjeojaiT912CX30MgdDr3iJS5TrTm\nZyEug4BIDS4N0KiQFSGQ/ED2Zw8Se1Ewwsy5TBJdpPQWVx5wTY2ERtkbKWRnI2XiTIzejCTR\nQZRAV8uMHd3Wvfue7t/GQaP6NH8dlzcYgf1fI1LgyAmKlHU0hPS5l7e6jvp6/fXzN2Tdv8WM\nRk65vEZyRprH5t3P4NRQuEtuc63E15dXJG+hHuHIqGKTnCvSbbePS/5Oz37/darQKNDFVCL5\nOdLmDvJE8gw5GBDf/n03tkov862avCdZpiQ7cJtJH5M3vplBn+9D7WSkf0r9nIw+/yj1z65D\nIz+SGu2S2n2PcxHJGPMSLU+AA/unl86XnqHc3y2bV083L9PDW7hbkv5H/Tct/LHr0Ki+RkXv\nbPB90znawTJS7falFiRtfkqkGXO1WTH+T5dI35hzDvd26SkrSma8WUm4WZ2T1HUS4qvwi31o\n5Ckt16jASMrbXksbrKFdLt7z+ONuz8247RKek5wW3h28jIx9m5exS5e3edteLG1NtSNYo/fu\ntraztbO9iJHQSBuMr80uRvK3D4g0TmXBSWoe6XzS1yp9N3RCp+H5ynMSyZBB6/j26FfbsBm1\nmxW8mz4Oox87BJZK3tTjYu9iNb7UL7sIjeprJGckkzl2j/9O+69nD7dHg4u9jk+ky6S41oG1\nIW8c59JJwKX3aY3pgLn/CokR6HzqxzoUtQxmMA7GYT6Mhmq8q1e7CI3qa7TeSPH3Q41RXeKw\nDNsYoraHWpid6Wrqbph30tpPO0fxpR3GIIyWS2vrhG+P96612fe45dsotnUpMGdIy+r5+cgq\nrteveRmNdtOo2luEJpEuy475g2RMV9akOO702MzIsN1Zw0gb7LAbM6Uxe+m6mHP0FPfHJDin\n+J6bWkvj25zjTFt8iKR1fCsWqYSX2OdIFmgkp1GZkVJfY552ZtB+TSMLXoMOwxQB5yrzLtJl\naXqx+tJEHJbU4BGFx9Tju6gMZsM/IdZmq9ttifxlEumi7cllHvy0xTkZmkRaUpT5IqJmXje8\nuokdGg07aFRkJF9jS6QpV7DCM8Zt0PZ8WmGcEeeGg77WsMg89a+3uhlbmeeSsc0Smstgrjiv\nZq1iDWcZq6W1X6Q5R7gsg35UjP/1HzGSeH2ERjtoVPju70ThRdv7aYfvV2/LqfJmqKQ1Gn9o\nZ+XbItJ82l560DajLWl1c+UUdv2Mf1mOp6l8CqYW4Lm3ObzadKtd9WrZzE07UgdHpOWIqHJe\n8vsIjYahukYlnyOl3uJp7OKs0mUYT6pT4jkqNmW1g9n4NlfPp35DcV2k220pn1fXZq9p6lrG\nMM2302j1s/uYaeub0XJ9a3qcxj8soZ+OxGmXhmVflkPptnSbH/lc/vh9hEY7aCT8NYrl5Hgx\nIzsERFqkvf+8OegaPS6Kp4nzUatXLzeM5sjO9aZIc/EwbXMc49LXcphok+5l2pFxbPMxuaRE\nl0U7bU/H42HZr61xd/n8Nb5HH40aaCT50Op8hrZFWnZjCaAWYW3KS4h0mUWaQmXU3ScTU6Rl\nstQyCm01rbFRvoxlmQI1kbTG405pc+CsnbYTlvJbo+7h83p179eh0V4ayRrpctFiakxDc9J6\nW3bIbjYM9g7fl43pUMuKjd0f9LhaobFWXLqcio1oztOg09F8GA1a8TRmvfNll80eBm1mlebr\nevV8EwmN9tJI9OnvSwA9DIO5jxdtrWWCWPbPVDAsxEWvG0wBbf2Gwdis0ZuVhJgxHobB23rQ\npl/jGn7cXa3hkpML2+lP5l+jQKM6GtUwkiOXHlVnz42WmkiPHVl2eDpBD0Zvng15ztYXY5ay\nxrdoZ4ikb8fX2pBU7/3iaKLtZ7WrpNw/64JGdTQSTe3m/bOWjfOqN4RmCIy7mmYmH+lErx2s\nuuAkadUEyrNKhyVXnwsHu6m2+tbIo1FPGlUxko2jkR5D78xj7Eq4weDU++eowVrR2GpIJN+F\naKjU10NATeMA3Ac0qq/RLkbyahSb7kJhc+ovMZHDK4ZqQit4g+yfw+7/ty4U7F5rfY6ERu00\namIkf/z9LZMizdhTWqZIZeX+bhJN/b1ujTwa9aTRLkYKqJTVLtEiUptZVVie0CMhkt7p1sij\nUU8ayb77O0SmRuUixWozq6aQRURKFlcWCY3616jhGSmrXaJJfmVKpDw10veJ4i3XiYRG/WvU\n8Bopq2GiRbQ2XBWqSKuxSiT/xrZGHo160ggjFZZvE6nmzYb7K4tf3r7cGjQK6SGn0RMbKbMq\nozxZ2sMZ6fP6eKzB88gdGqX18G8sP/wYSUIkb6mcSFn8Vi8fw/Dxon6jUQONMNKuItVL7T6n\n19ldlZPcoVGoVE4juVcWe2Iejv0eIlW5kN0okr6xgshn8K7eHgtv00vA0ajTM1L6lcUhnkik\n5BwWb7lOpBz+TH+E4iPyEn2fOMHQoxFG8tTJi+SPva/Qv7H8yOfwosabDJ/qBY321wgjrRYp\n1dZXuE6kLHWUu4RGqbZyGq03kv0tMp864eAvI1fxhvGuIpUqM20IqVRbpPzA56njMxIaJdrK\nabT5jFTab/rleP4Su0DFalX5fzKWU4uJQm9DMWJnpNAqRRVoFAUj5S9jpETJmTXCSPnLGClR\ncmaNMFL+ctdG+j19Dvvl3rULgEZZS3mUtA//pZ2CfhGpjpHmz5H+uZ8jBUCjrKU8amga7ReR\n6gT9v+XJhsCLix3QKGspD4yUv9y1kT7V9bFwVcE3rlqgUdZSHhgpf7lrI01Pf//yPP0dAI2y\nlvKoZSTYm3/h7yNBfTDS03D/huwv3zdkA7xni//1elXXP/+KhvP5vY7KWec+7j8fuf0uoy4a\nlbazf1+Uegn95Y61YKTT8pl9D/bv+KTRW0Hv0zrJA3b6Zq//b6S5zedRF41K29nXou3lgpHO\nys8BnNfy66p+znP/KZV92hj+qet/X8PXm1KpM8ZV/flORr9b5hliHnXRqLSd/atevof0z/fX\n37eAkU7K/USQ1/R1PMb/y/6odximvx74lvpU6226OfLqfrPXRRt1yaj0nX15GO+jYGdywEjn\n5P07t8k10q/pEM//PP5juhn/NS2EeJlOKJ/uN3sd9FEXjMrY2Wkhf2eywEjnRH2fMYoPpfwV\nMhM1o8+M+/beUadHZayGkUCQ35/lh9J/2Q8fDb+Tl0YTmpF+pbv1jDpjVMZqpHYgTJmRPl/V\nNfvW+nfedb85/ZpcY87Qcm8ims2yRzWv9nG/2fD5UnDnJKt/0d7gSJQY6f3no978j6iUGu9q\nX1OH6+t0H/rPCiMVjGpZ7e99ZME/Ab+SSkYKdOt/oVf0LV/5/SQrVXLVrEapFrH6sbjkOfp6\nFBnp9++r+pU9iX8fqD8PLH29J4/zn7vY35nX5+/cu/GGkfJHNa/2+fvxOVJ2mppHHT2jR3JW\nYcbBXlr5OIRjq2Y1SrWI1Y/FqUHsRPk1Uv7zsNOdg7fkB58f46lrXWqXPSrt46f3b29/vqdu\nJxZSRU4VPVRzCjOMFK30tHgMKnEMayMPnVSXblL1wR5i/e9I8VnxPf+rTsvLwZK3EKaLqYyb\nDY+uV41q+fjpfVpN9NmGfVO7YJWy/ytvpCHHSEPSSEPCSOnNHNdI+Wu8LGHMXeVf5mPrnreN\n5WxianRdvkksekrqwUjuBUPSSKuvkTKNlLpGio1g2kyowRmM9KfcSLkfPW000op3W2Sxu5Ei\nV0/6/+JGip0U9jkjRUbw2IQKTQbd+Cj7SLoud6jzsq+fC5e/0yqJj2vmznPvRzunlrxRLauN\nBQc/I8Vn6eV/K7O3aKV0apfal1gS24GPso2k3aHO/Rb713Rwv6ZWmfr8l2vS5WKnaFTLauOT\nSMe+RsoqN98OGmp9TCN15KNsI32qx42uPwVz+Jt6eTzTnVrl38/zO8PX+zX3WYhp1IWj0h8a\n/8y6MV/GvkYKHlpu3pvsvaKR4o1yjBTK/HryUf5Fwt/xK0O/Cr58+5r7fd3/cr+4NLJ8H6Jo\nVM5qwp/I7mok/6nmWT+QDQxNqfQpdx/yh/D487S5ed2Djz/fx3j6EaGChg+WUReNylqtYIOZ\n/ct2B3BOMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAj\nAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIcGwjjS+1OvZOPDkn0ejQ+xd7FSP0\nwVk0OvL+qennkXfiyTmNRkfeP+0l2+Nfh1BTJvEoUfpLT41fnnqowWk06n18MRaRHjPe+Gte\nVPNU6P5y6qEKp9Go8+FFUfqCFm7voko1hRqcRqPOhxelSKQ5X3Dqu3ih/bNyGo06H14UNf3M\nEinWFGpxGo0OMMQwakgp48hyyLThyJxFo86Hl0D7sG+at4zsQMsX7DtCTj3U4SQa9T4+gEOA\nkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAA\nRgIQACMBCICRAATASAACYCQAAf4fLc//uctPLscAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "par(mfrow = c(3,2))\n",
    "manhattan(gwasResults)\n",
    "manhattan(gwasResults,suggestiveline = F)\n",
    "manhattan(gwasResults,ylim = c(0,10))\n",
    "manhattan(gwasResults,genomewideline = 10,ylim = c(0,10))\n",
    "manhattan(gwasResults,genomewideline = 10,ylim = c(0,10),cex = 0.5)\n",
    "manhattan(gwasResults,genomewideline = 10,ylim = c(0,10),cex = 0.5,cex.axis = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (x, chr = \"CHR\", bp = \"BP\", p = \"P\", snp = \"SNP\", col = c(\"gray10\", \n",
       "<span style=white-space:pre-wrap>    \"gray60\"), chrlabs = NULL, suggestiveline = -log10(1e-05), </span>\n",
       "<span style=white-space:pre-wrap>    genomewideline = -log10(5e-08), highlight = NULL, logp = TRUE, </span>\n",
       "<span style=white-space:pre-wrap>    annotatePval = NULL, annotateTop = TRUE, ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    CHR = BP = P = index = NULL</span>\n",
       "<span style=white-space:pre-wrap>    if (!(chr %in% names(x))) </span>\n",
       "<span style=white-space:pre-wrap>        stop(paste(\"Column\", chr, \"not found!\"))</span>\n",
       "<span style=white-space:pre-wrap>    if (!(bp %in% names(x))) </span>\n",
       "<span style=white-space:pre-wrap>        stop(paste(\"Column\", bp, \"not found!\"))</span>\n",
       "<span style=white-space:pre-wrap>    if (!(p %in% names(x))) </span>\n",
       "<span style=white-space:pre-wrap>        stop(paste(\"Column\", p, \"not found!\"))</span>\n",
       "<span style=white-space:pre-wrap>    if (!(snp %in% names(x))) </span>\n",
       "<span style=white-space:pre-wrap>        warning(paste(\"No SNP column found. OK unless you're trying to highlight.\"))</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.numeric(x[[chr]])) </span>\n",
       "<span style=white-space:pre-wrap>        stop(paste(chr, \"column should be numeric. Do you have 'X', 'Y', 'MT', etc? If so change to numbers and try again.\"))</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.numeric(x[[bp]])) </span>\n",
       "<span style=white-space:pre-wrap>        stop(paste(bp, \"column should be numeric.\"))</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.numeric(x[[p]])) </span>\n",
       "<span style=white-space:pre-wrap>        stop(paste(p, \"column should be numeric.\"))</span>\n",
       "<span style=white-space:pre-wrap>    d = data.frame(CHR = x[[chr]], BP = x[[bp]], P = x[[p]])</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(x[[snp]])) </span>\n",
       "<span style=white-space:pre-wrap>        d = transform(d, SNP = x[[snp]])</span>\n",
       "<span style=white-space:pre-wrap>    d &lt;- subset(d, (is.numeric(CHR) &amp; is.numeric(BP) &amp; is.numeric(P)))</span>\n",
       "<span style=white-space:pre-wrap>    d &lt;- d[order(d$CHR, d$BP), ]</span>\n",
       "<span style=white-space:pre-wrap>    if (logp) {</span>\n",
       "<span style=white-space:pre-wrap>        d$logp &lt;- -log10(d$P)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        d$logp &lt;- d$P</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    d$pos = NA</span>\n",
       "<span style=white-space:pre-wrap>    d$index = NA</span>\n",
       "<span style=white-space:pre-wrap>    ind = 0</span>\n",
       "<span style=white-space:pre-wrap>    for (i in unique(d$CHR)) {</span>\n",
       "<span style=white-space:pre-wrap>        ind = ind + 1</span>\n",
       "<span style=white-space:pre-wrap>        d[d$CHR == i, ]$index = ind</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    nchr = length(unique(d$CHR))</span>\n",
       "<span style=white-space:pre-wrap>    if (nchr == 1) {</span>\n",
       "<span style=white-space:pre-wrap>        d$pos = d$BP</span>\n",
       "<span style=white-space:pre-wrap>        ticks = floor(length(d$pos))/2 + 1</span>\n",
       "<span style=white-space:pre-wrap>        xlabel = paste(\"Chromosome\", unique(d$CHR), \"position\")</span>\n",
       "<span style=white-space:pre-wrap>        labs = ticks</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        lastbase = 0</span>\n",
       "<span style=white-space:pre-wrap>        ticks = NULL</span>\n",
       "<span style=white-space:pre-wrap>        for (i in unique(d$index)) {</span>\n",
       "<span style=white-space:pre-wrap>            if (i == 1) {</span>\n",
       "<span style=white-space:pre-wrap>                d[d$index == i, ]$pos = d[d$index == i, ]$BP</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            else {</span>\n",
       "<span style=white-space:pre-wrap>                lastbase = lastbase + tail(subset(d, index == </span>\n",
       "<span style=white-space:pre-wrap>                  i - 1)$BP, 1)</span>\n",
       "<span style=white-space:pre-wrap>                d[d$index == i, ]$pos = d[d$index == i, ]$BP + </span>\n",
       "<span style=white-space:pre-wrap>                  lastbase</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            ticks = c(ticks, (min(d[d$index == i, ]$pos) + max(d[d$index == </span>\n",
       "<span style=white-space:pre-wrap>                i, ]$pos))/2 + 1)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        xlabel = \"Chromosome\"</span>\n",
       "<span style=white-space:pre-wrap>        labs &lt;- unique(d$CHR)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    xmax = ceiling(max(d$pos) * 1.03)</span>\n",
       "<span style=white-space:pre-wrap>    xmin = floor(max(d$pos) * -0.03)</span>\n",
       "<span style=white-space:pre-wrap>    def_args &lt;- list(xaxt = \"n\", bty = \"n\", xaxs = \"i\", yaxs = \"i\", </span>\n",
       "<span style=white-space:pre-wrap>        las = 1, pch = 20, xlim = c(xmin, xmax), ylim = c(0, </span>\n",
       "<span style=white-space:pre-wrap>            ceiling(max(d$logp))), xlab = xlabel, ylab = expression(-log[10](italic(p))))</span>\n",
       "<span style=white-space:pre-wrap>    dotargs &lt;- list(...)</span>\n",
       "<span style=white-space:pre-wrap>    do.call(\"plot\", c(NA, dotargs, def_args[!names(def_args) %in% </span>\n",
       "<span style=white-space:pre-wrap>        names(dotargs)]))</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(chrlabs)) {</span>\n",
       "<span style=white-space:pre-wrap>        if (is.character(chrlabs)) {</span>\n",
       "<span style=white-space:pre-wrap>            if (length(chrlabs) == length(labs)) {</span>\n",
       "<span style=white-space:pre-wrap>                labs &lt;- chrlabs</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            else {</span>\n",
       "<span style=white-space:pre-wrap>                warning(\"You're trying to specify chromosome labels but the number of labels != number of chromosomes.\")</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            warning(\"If you're trying to specify chromosome labels, chrlabs must be a character vector\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (nchr == 1) {</span>\n",
       "<span style=white-space:pre-wrap>        axis(1, ...)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        axis(1, at = ticks, labels = labs, ...)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    col = rep(col, max(d$CHR))</span>\n",
       "<span style=white-space:pre-wrap>    if (nchr == 1) {</span>\n",
       "<span style=white-space:pre-wrap>        with(d, points(pos, logp, pch = 20, col = col[1], ...))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        icol = 1</span>\n",
       "<span style=white-space:pre-wrap>        for (i in unique(d$index)) {</span>\n",
       "<span style=white-space:pre-wrap>            with(d[d$index == unique(d$index)[i], ], points(pos, </span>\n",
       "<span style=white-space:pre-wrap>                logp, col = col[icol], pch = 20, ...))</span>\n",
       "<span style=white-space:pre-wrap>            icol = icol + 1</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (suggestiveline) </span>\n",
       "<span style=white-space:pre-wrap>        abline(h = suggestiveline, col = \"blue\")</span>\n",
       "<span style=white-space:pre-wrap>    if (genomewideline) </span>\n",
       "<span style=white-space:pre-wrap>        abline(h = genomewideline, col = \"red\")</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(highlight)) {</span>\n",
       "<span style=white-space:pre-wrap>        if (any(!(highlight %in% d$SNP))) </span>\n",
       "<span style=white-space:pre-wrap>            warning(\"You're trying to highlight SNPs that don't exist in your results.\")</span>\n",
       "<span style=white-space:pre-wrap>        d.highlight = d[which(d$SNP %in% highlight), ]</span>\n",
       "<span style=white-space:pre-wrap>        with(d.highlight, points(pos, logp, col = \"green3\", pch = 20, </span>\n",
       "<span style=white-space:pre-wrap>            ...))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(annotatePval)) {</span>\n",
       "<span style=white-space:pre-wrap>        topHits = subset(d, P &lt;= annotatePval)</span>\n",
       "<span style=white-space:pre-wrap>        par(xpd = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        if (annotateTop == FALSE) {</span>\n",
       "<span style=white-space:pre-wrap>            with(subset(d, P &lt;= annotatePval), textxy(pos, -log10(P), </span>\n",
       "<span style=white-space:pre-wrap>                offset = 0.625, labs = topHits$SNP, cex = 0.45), </span>\n",
       "<span style=white-space:pre-wrap>                ...)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            topHits &lt;- topHits[order(topHits$P), ]</span>\n",
       "<span style=white-space:pre-wrap>            topSNPs &lt;- NULL</span>\n",
       "<span style=white-space:pre-wrap>            for (i in unique(topHits$CHR)) {</span>\n",
       "<span style=white-space:pre-wrap>                chrSNPs &lt;- topHits[topHits$CHR == i, ]</span>\n",
       "<span style=white-space:pre-wrap>                topSNPs &lt;- rbind(topSNPs, chrSNPs[1, ])</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            textxy(topSNPs$pos, -log10(topSNPs$P), offset = 0.625, </span>\n",
       "<span style=white-space:pre-wrap>                labs = topSNPs$SNP, cex = 0.5, ...)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    par(xpd = FALSE)</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (x, chr = \"CHR\", bp = \"BP\", p = \"P\", snp = \"SNP\", col = c(\"gray10\", \n",
       "    \"gray60\"), chrlabs = NULL, suggestiveline = -log10(1e-05), \n",
       "    genomewideline = -log10(5e-08), highlight = NULL, logp = TRUE, \n",
       "    annotatePval = NULL, annotateTop = TRUE, ...) \n",
       "\\{\n",
       "    CHR = BP = P = index = NULL\n",
       "    if (!(chr \\%in\\% names(x))) \n",
       "        stop(paste(\"Column\", chr, \"not found!\"))\n",
       "    if (!(bp \\%in\\% names(x))) \n",
       "        stop(paste(\"Column\", bp, \"not found!\"))\n",
       "    if (!(p \\%in\\% names(x))) \n",
       "        stop(paste(\"Column\", p, \"not found!\"))\n",
       "    if (!(snp \\%in\\% names(x))) \n",
       "        warning(paste(\"No SNP column found. OK unless you're trying to highlight.\"))\n",
       "    if (!is.numeric(x{[}{[}chr{]}{]})) \n",
       "        stop(paste(chr, \"column should be numeric. Do you have 'X', 'Y', 'MT', etc? If so change to numbers and try again.\"))\n",
       "    if (!is.numeric(x{[}{[}bp{]}{]})) \n",
       "        stop(paste(bp, \"column should be numeric.\"))\n",
       "    if (!is.numeric(x{[}{[}p{]}{]})) \n",
       "        stop(paste(p, \"column should be numeric.\"))\n",
       "    d = data.frame(CHR = x{[}{[}chr{]}{]}, BP = x{[}{[}bp{]}{]}, P = x{[}{[}p{]}{]})\n",
       "    if (!is.null(x{[}{[}snp{]}{]})) \n",
       "        d = transform(d, SNP = x{[}{[}snp{]}{]})\n",
       "    d <- subset(d, (is.numeric(CHR) \\& is.numeric(BP) \\& is.numeric(P)))\n",
       "    d <- d{[}order(d\\$CHR, d\\$BP), {]}\n",
       "    if (logp) \\{\n",
       "        d\\$logp <- -log10(d\\$P)\n",
       "    \\}\n",
       "    else \\{\n",
       "        d\\$logp <- d\\$P\n",
       "    \\}\n",
       "    d\\$pos = NA\n",
       "    d\\$index = NA\n",
       "    ind = 0\n",
       "    for (i in unique(d\\$CHR)) \\{\n",
       "        ind = ind + 1\n",
       "        d{[}d\\$CHR == i, {]}\\$index = ind\n",
       "    \\}\n",
       "    nchr = length(unique(d\\$CHR))\n",
       "    if (nchr == 1) \\{\n",
       "        d\\$pos = d\\$BP\n",
       "        ticks = floor(length(d\\$pos))/2 + 1\n",
       "        xlabel = paste(\"Chromosome\", unique(d\\$CHR), \"position\")\n",
       "        labs = ticks\n",
       "    \\}\n",
       "    else \\{\n",
       "        lastbase = 0\n",
       "        ticks = NULL\n",
       "        for (i in unique(d\\$index)) \\{\n",
       "            if (i == 1) \\{\n",
       "                d{[}d\\$index == i, {]}\\$pos = d{[}d\\$index == i, {]}\\$BP\n",
       "            \\}\n",
       "            else \\{\n",
       "                lastbase = lastbase + tail(subset(d, index == \n",
       "                  i - 1)\\$BP, 1)\n",
       "                d{[}d\\$index == i, {]}\\$pos = d{[}d\\$index == i, {]}\\$BP + \n",
       "                  lastbase\n",
       "            \\}\n",
       "            ticks = c(ticks, (min(d{[}d\\$index == i, {]}\\$pos) + max(d{[}d\\$index == \n",
       "                i, {]}\\$pos))/2 + 1)\n",
       "        \\}\n",
       "        xlabel = \"Chromosome\"\n",
       "        labs <- unique(d\\$CHR)\n",
       "    \\}\n",
       "    xmax = ceiling(max(d\\$pos) * 1.03)\n",
       "    xmin = floor(max(d\\$pos) * -0.03)\n",
       "    def\\_args <- list(xaxt = \"n\", bty = \"n\", xaxs = \"i\", yaxs = \"i\", \n",
       "        las = 1, pch = 20, xlim = c(xmin, xmax), ylim = c(0, \n",
       "            ceiling(max(d\\$logp))), xlab = xlabel, ylab = expression(-log{[}10{]}(italic(p))))\n",
       "    dotargs <- list(...)\n",
       "    do.call(\"plot\", c(NA, dotargs, def\\_args{[}!names(def\\_args) \\%in\\% \n",
       "        names(dotargs){]}))\n",
       "    if (!is.null(chrlabs)) \\{\n",
       "        if (is.character(chrlabs)) \\{\n",
       "            if (length(chrlabs) == length(labs)) \\{\n",
       "                labs <- chrlabs\n",
       "            \\}\n",
       "            else \\{\n",
       "                warning(\"You're trying to specify chromosome labels but the number of labels != number of chromosomes.\")\n",
       "            \\}\n",
       "        \\}\n",
       "        else \\{\n",
       "            warning(\"If you're trying to specify chromosome labels, chrlabs must be a character vector\")\n",
       "        \\}\n",
       "    \\}\n",
       "    if (nchr == 1) \\{\n",
       "        axis(1, ...)\n",
       "    \\}\n",
       "    else \\{\n",
       "        axis(1, at = ticks, labels = labs, ...)\n",
       "    \\}\n",
       "    col = rep(col, max(d\\$CHR))\n",
       "    if (nchr == 1) \\{\n",
       "        with(d, points(pos, logp, pch = 20, col = col{[}1{]}, ...))\n",
       "    \\}\n",
       "    else \\{\n",
       "        icol = 1\n",
       "        for (i in unique(d\\$index)) \\{\n",
       "            with(d{[}d\\$index == unique(d\\$index){[}i{]}, {]}, points(pos, \n",
       "                logp, col = col{[}icol{]}, pch = 20, ...))\n",
       "            icol = icol + 1\n",
       "        \\}\n",
       "    \\}\n",
       "    if (suggestiveline) \n",
       "        abline(h = suggestiveline, col = \"blue\")\n",
       "    if (genomewideline) \n",
       "        abline(h = genomewideline, col = \"red\")\n",
       "    if (!is.null(highlight)) \\{\n",
       "        if (any(!(highlight \\%in\\% d\\$SNP))) \n",
       "            warning(\"You're trying to highlight SNPs that don't exist in your results.\")\n",
       "        d.highlight = d{[}which(d\\$SNP \\%in\\% highlight), {]}\n",
       "        with(d.highlight, points(pos, logp, col = \"green3\", pch = 20, \n",
       "            ...))\n",
       "    \\}\n",
       "    if (!is.null(annotatePval)) \\{\n",
       "        topHits = subset(d, P <= annotatePval)\n",
       "        par(xpd = TRUE)\n",
       "        if (annotateTop == FALSE) \\{\n",
       "            with(subset(d, P <= annotatePval), textxy(pos, -log10(P), \n",
       "                offset = 0.625, labs = topHits\\$SNP, cex = 0.45), \n",
       "                ...)\n",
       "        \\}\n",
       "        else \\{\n",
       "            topHits <- topHits{[}order(topHits\\$P), {]}\n",
       "            topSNPs <- NULL\n",
       "            for (i in unique(topHits\\$CHR)) \\{\n",
       "                chrSNPs <- topHits{[}topHits\\$CHR == i, {]}\n",
       "                topSNPs <- rbind(topSNPs, chrSNPs{[}1, {]})\n",
       "            \\}\n",
       "            textxy(topSNPs\\$pos, -log10(topSNPs\\$P), offset = 0.625, \n",
       "                labs = topSNPs\\$SNP, cex = 0.5, ...)\n",
       "        \\}\n",
       "    \\}\n",
       "    par(xpd = FALSE)\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (x, chr = \"CHR\", bp = \"BP\", p = \"P\", snp = \"SNP\", col = c(\"gray10\", \n",
       "    \"gray60\"), chrlabs = NULL, suggestiveline = -log10(1e-05), \n",
       "    genomewideline = -log10(5e-08), highlight = NULL, logp = TRUE, \n",
       "    annotatePval = NULL, annotateTop = TRUE, ...) \n",
       "{\n",
       "    CHR = BP = P = index = NULL\n",
       "    if (!(chr %in% names(x))) \n",
       "        stop(paste(\"Column\", chr, \"not found!\"))\n",
       "    if (!(bp %in% names(x))) \n",
       "        stop(paste(\"Column\", bp, \"not found!\"))\n",
       "    if (!(p %in% names(x))) \n",
       "        stop(paste(\"Column\", p, \"not found!\"))\n",
       "    if (!(snp %in% names(x))) \n",
       "        warning(paste(\"No SNP column found. OK unless you're trying to highlight.\"))\n",
       "    if (!is.numeric(x[[chr]])) \n",
       "        stop(paste(chr, \"column should be numeric. Do you have 'X', 'Y', 'MT', etc? If so change to numbers and try again.\"))\n",
       "    if (!is.numeric(x[[bp]])) \n",
       "        stop(paste(bp, \"column should be numeric.\"))\n",
       "    if (!is.numeric(x[[p]])) \n",
       "        stop(paste(p, \"column should be numeric.\"))\n",
       "    d = data.frame(CHR = x[[chr]], BP = x[[bp]], P = x[[p]])\n",
       "    if (!is.null(x[[snp]])) \n",
       "        d = transform(d, SNP = x[[snp]])\n",
       "    d <- subset(d, (is.numeric(CHR) & is.numeric(BP) & is.numeric(P)))\n",
       "    d <- d[order(d$CHR, d$BP), ]\n",
       "    if (logp) {\n",
       "        d$logp <- -log10(d$P)\n",
       "    }\n",
       "    else {\n",
       "        d$logp <- d$P\n",
       "    }\n",
       "    d$pos = NA\n",
       "    d$index = NA\n",
       "    ind = 0\n",
       "    for (i in unique(d$CHR)) {\n",
       "        ind = ind + 1\n",
       "        d[d$CHR == i, ]$index = ind\n",
       "    }\n",
       "    nchr = length(unique(d$CHR))\n",
       "    if (nchr == 1) {\n",
       "        d$pos = d$BP\n",
       "        ticks = floor(length(d$pos))/2 + 1\n",
       "        xlabel = paste(\"Chromosome\", unique(d$CHR), \"position\")\n",
       "        labs = ticks\n",
       "    }\n",
       "    else {\n",
       "        lastbase = 0\n",
       "        ticks = NULL\n",
       "        for (i in unique(d$index)) {\n",
       "            if (i == 1) {\n",
       "                d[d$index == i, ]$pos = d[d$index == i, ]$BP\n",
       "            }\n",
       "            else {\n",
       "                lastbase = lastbase + tail(subset(d, index == \n",
       "                  i - 1)$BP, 1)\n",
       "                d[d$index == i, ]$pos = d[d$index == i, ]$BP + \n",
       "                  lastbase\n",
       "            }\n",
       "            ticks = c(ticks, (min(d[d$index == i, ]$pos) + max(d[d$index == \n",
       "                i, ]$pos))/2 + 1)\n",
       "        }\n",
       "        xlabel = \"Chromosome\"\n",
       "        labs <- unique(d$CHR)\n",
       "    }\n",
       "    xmax = ceiling(max(d$pos) * 1.03)\n",
       "    xmin = floor(max(d$pos) * -0.03)\n",
       "    def_args <- list(xaxt = \"n\", bty = \"n\", xaxs = \"i\", yaxs = \"i\", \n",
       "        las = 1, pch = 20, xlim = c(xmin, xmax), ylim = c(0, \n",
       "            ceiling(max(d$logp))), xlab = xlabel, ylab = expression(-log[10](italic(p))))\n",
       "    dotargs <- list(...)\n",
       "    do.call(\"plot\", c(NA, dotargs, def_args[!names(def_args) %in% \n",
       "        names(dotargs)]))\n",
       "    if (!is.null(chrlabs)) {\n",
       "        if (is.character(chrlabs)) {\n",
       "            if (length(chrlabs) == length(labs)) {\n",
       "                labs <- chrlabs\n",
       "            }\n",
       "            else {\n",
       "                warning(\"You're trying to specify chromosome labels but the number of labels != number of chromosomes.\")\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            warning(\"If you're trying to specify chromosome labels, chrlabs must be a character vector\")\n",
       "        }\n",
       "    }\n",
       "    if (nchr == 1) {\n",
       "        axis(1, ...)\n",
       "    }\n",
       "    else {\n",
       "        axis(1, at = ticks, labels = labs, ...)\n",
       "    }\n",
       "    col = rep(col, max(d$CHR))\n",
       "    if (nchr == 1) {\n",
       "        with(d, points(pos, logp, pch = 20, col = col[1], ...))\n",
       "    }\n",
       "    else {\n",
       "        icol = 1\n",
       "        for (i in unique(d$index)) {\n",
       "            with(d[d$index == unique(d$index)[i], ], points(pos, \n",
       "                logp, col = col[icol], pch = 20, ...))\n",
       "            icol = icol + 1\n",
       "        }\n",
       "    }\n",
       "    if (suggestiveline) \n",
       "        abline(h = suggestiveline, col = \"blue\")\n",
       "    if (genomewideline) \n",
       "        abline(h = genomewideline, col = \"red\")\n",
       "    if (!is.null(highlight)) {\n",
       "        if (any(!(highlight %in% d$SNP))) \n",
       "            warning(\"You're trying to highlight SNPs that don't exist in your results.\")\n",
       "        d.highlight = d[which(d$SNP %in% highlight), ]\n",
       "        with(d.highlight, points(pos, logp, col = \"green3\", pch = 20, \n",
       "            ...))\n",
       "    }\n",
       "    if (!is.null(annotatePval)) {\n",
       "        topHits = subset(d, P <= annotatePval)\n",
       "        par(xpd = TRUE)\n",
       "        if (annotateTop == FALSE) {\n",
       "            with(subset(d, P <= annotatePval), textxy(pos, -log10(P), \n",
       "                offset = 0.625, labs = topHits$SNP, cex = 0.45), \n",
       "                ...)\n",
       "        }\n",
       "        else {\n",
       "            topHits <- topHits[order(topHits$P), ]\n",
       "            topSNPs <- NULL\n",
       "            for (i in unique(topHits$CHR)) {\n",
       "                chrSNPs <- topHits[topHits$CHR == i, ]\n",
       "                topSNPs <- rbind(topSNPs, chrSNPs[1, ])\n",
       "            }\n",
       "            textxy(topSNPs$pos, -log10(topSNPs$P), offset = 0.625, \n",
       "                labs = topSNPs$SNP, cex = 0.5, ...)\n",
       "        }\n",
       "    }\n",
       "    par(xpd = FALSE)\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (x, chr = \"CHR\", bp = \"BP\", p = \"P\", snp = \"SNP\", col = c(\"gray10\", \n",
       "    \"gray60\"), chrlabs = NULL, suggestiveline = -log10(1e-05), \n",
       "    genomewideline = -log10(5e-08), highlight = NULL, logp = TRUE, \n",
       "    annotatePval = NULL, annotateTop = TRUE, ...) \n",
       "{\n",
       "    CHR = BP = P = index = NULL\n",
       "    if (!(chr %in% names(x))) \n",
       "        stop(paste(\"Column\", chr, \"not found!\"))\n",
       "    if (!(bp %in% names(x))) \n",
       "        stop(paste(\"Column\", bp, \"not found!\"))\n",
       "    if (!(p %in% names(x))) \n",
       "        stop(paste(\"Column\", p, \"not found!\"))\n",
       "    if (!(snp %in% names(x))) \n",
       "        warning(paste(\"No SNP column found. OK unless you're trying to highlight.\"))\n",
       "    if (!is.numeric(x[[chr]])) \n",
       "        stop(paste(chr, \"column should be numeric. Do you have 'X', 'Y', 'MT', etc? If so change to numbers and try again.\"))\n",
       "    if (!is.numeric(x[[bp]])) \n",
       "        stop(paste(bp, \"column should be numeric.\"))\n",
       "    if (!is.numeric(x[[p]])) \n",
       "        stop(paste(p, \"column should be numeric.\"))\n",
       "    d = data.frame(CHR = x[[chr]], BP = x[[bp]], P = x[[p]])\n",
       "    if (!is.null(x[[snp]])) \n",
       "        d = transform(d, SNP = x[[snp]])\n",
       "    d <- subset(d, (is.numeric(CHR) & is.numeric(BP) & is.numeric(P)))\n",
       "    d <- d[order(d$CHR, d$BP), ]\n",
       "    if (logp) {\n",
       "        d$logp <- -log10(d$P)\n",
       "    }\n",
       "    else {\n",
       "        d$logp <- d$P\n",
       "    }\n",
       "    d$pos = NA\n",
       "    d$index = NA\n",
       "    ind = 0\n",
       "    for (i in unique(d$CHR)) {\n",
       "        ind = ind + 1\n",
       "        d[d$CHR == i, ]$index = ind\n",
       "    }\n",
       "    nchr = length(unique(d$CHR))\n",
       "    if (nchr == 1) {\n",
       "        d$pos = d$BP\n",
       "        ticks = floor(length(d$pos))/2 + 1\n",
       "        xlabel = paste(\"Chromosome\", unique(d$CHR), \"position\")\n",
       "        labs = ticks\n",
       "    }\n",
       "    else {\n",
       "        lastbase = 0\n",
       "        ticks = NULL\n",
       "        for (i in unique(d$index)) {\n",
       "            if (i == 1) {\n",
       "                d[d$index == i, ]$pos = d[d$index == i, ]$BP\n",
       "            }\n",
       "            else {\n",
       "                lastbase = lastbase + tail(subset(d, index == \n",
       "                  i - 1)$BP, 1)\n",
       "                d[d$index == i, ]$pos = d[d$index == i, ]$BP + \n",
       "                  lastbase\n",
       "            }\n",
       "            ticks = c(ticks, (min(d[d$index == i, ]$pos) + max(d[d$index == \n",
       "                i, ]$pos))/2 + 1)\n",
       "        }\n",
       "        xlabel = \"Chromosome\"\n",
       "        labs <- unique(d$CHR)\n",
       "    }\n",
       "    xmax = ceiling(max(d$pos) * 1.03)\n",
       "    xmin = floor(max(d$pos) * -0.03)\n",
       "    def_args <- list(xaxt = \"n\", bty = \"n\", xaxs = \"i\", yaxs = \"i\", \n",
       "        las = 1, pch = 20, xlim = c(xmin, xmax), ylim = c(0, \n",
       "            ceiling(max(d$logp))), xlab = xlabel, ylab = expression(-log[10](italic(p))))\n",
       "    dotargs <- list(...)\n",
       "    do.call(\"plot\", c(NA, dotargs, def_args[!names(def_args) %in% \n",
       "        names(dotargs)]))\n",
       "    if (!is.null(chrlabs)) {\n",
       "        if (is.character(chrlabs)) {\n",
       "            if (length(chrlabs) == length(labs)) {\n",
       "                labs <- chrlabs\n",
       "            }\n",
       "            else {\n",
       "                warning(\"You're trying to specify chromosome labels but the number of labels != number of chromosomes.\")\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            warning(\"If you're trying to specify chromosome labels, chrlabs must be a character vector\")\n",
       "        }\n",
       "    }\n",
       "    if (nchr == 1) {\n",
       "        axis(1, ...)\n",
       "    }\n",
       "    else {\n",
       "        axis(1, at = ticks, labels = labs, ...)\n",
       "    }\n",
       "    col = rep(col, max(d$CHR))\n",
       "    if (nchr == 1) {\n",
       "        with(d, points(pos, logp, pch = 20, col = col[1], ...))\n",
       "    }\n",
       "    else {\n",
       "        icol = 1\n",
       "        for (i in unique(d$index)) {\n",
       "            with(d[d$index == unique(d$index)[i], ], points(pos, \n",
       "                logp, col = col[icol], pch = 20, ...))\n",
       "            icol = icol + 1\n",
       "        }\n",
       "    }\n",
       "    if (suggestiveline) \n",
       "        abline(h = suggestiveline, col = \"blue\")\n",
       "    if (genomewideline) \n",
       "        abline(h = genomewideline, col = \"red\")\n",
       "    if (!is.null(highlight)) {\n",
       "        if (any(!(highlight %in% d$SNP))) \n",
       "            warning(\"You're trying to highlight SNPs that don't exist in your results.\")\n",
       "        d.highlight = d[which(d$SNP %in% highlight), ]\n",
       "        with(d.highlight, points(pos, logp, col = \"green3\", pch = 20, \n",
       "            ...))\n",
       "    }\n",
       "    if (!is.null(annotatePval)) {\n",
       "        topHits = subset(d, P <= annotatePval)\n",
       "        par(xpd = TRUE)\n",
       "        if (annotateTop == FALSE) {\n",
       "            with(subset(d, P <= annotatePval), textxy(pos, -log10(P), \n",
       "                offset = 0.625, labs = topHits$SNP, cex = 0.45), \n",
       "                ...)\n",
       "        }\n",
       "        else {\n",
       "            topHits <- topHits[order(topHits$P), ]\n",
       "            topSNPs <- NULL\n",
       "            for (i in unique(topHits$CHR)) {\n",
       "                chrSNPs <- topHits[topHits$CHR == i, ]\n",
       "                topSNPs <- rbind(topSNPs, chrSNPs[1, ])\n",
       "            }\n",
       "            textxy(topSNPs$pos, -log10(topSNPs$P), offset = 0.625, \n",
       "                labs = topSNPs$SNP, cex = 0.5, ...)\n",
       "        }\n",
       "    }\n",
       "    par(xpd = FALSE)\n",
       "}\n",
       "<bytecode: 0x0000000032fc8978>\n",
       "<environment: namespace:qqman>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "manhattan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(heatmap3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
